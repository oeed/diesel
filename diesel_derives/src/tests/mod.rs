trait ProcMacroFn {
    type Input: Clone;

    fn call(&self, input: Self::Input) -> proc_macro2::TokenStream;
}

impl ProcMacroFn for &dyn Fn(proc_macro2::TokenStream) -> proc_macro2::TokenStream {
    type Input = proc_macro2::TokenStream;

    fn call(&self, input: Self::Input) -> proc_macro2::TokenStream {
        (self)(input)
    }
}

impl ProcMacroFn
    for &dyn Fn(proc_macro2::TokenStream, proc_macro2::TokenStream) -> proc_macro2::TokenStream
{
    type Input = (proc_macro2::TokenStream, proc_macro2::TokenStream);

    fn call(&self, (attrs, input): Self::Input) -> proc_macro2::TokenStream {
        (self)(attrs, input)
    }
}

trait AttributeFormatter<I> {
    fn format(&self, input: I) -> String;
}

impl AttributeFormatter<proc_macro2::TokenStream> for syn::Attribute {
    fn format(&self, input: proc_macro2::TokenStream) -> String {
        format_input(quote::quote! {
            #self
            #input
        })
    }
}

fn derive(attr: syn::Attribute) -> syn::Attribute {
    attr
}

struct FunctionMacro(syn::Ident);

impl AttributeFormatter<proc_macro2::TokenStream> for FunctionMacro {
    fn format(&self, input: proc_macro2::TokenStream) -> String {
        let name = &self.0;
        format_input(quote::quote! {
            #name! {
                #input
            }
        })
    }
}

struct AttributeMacro(syn::Path);

impl AttributeFormatter<(proc_macro2::TokenStream, proc_macro2::TokenStream)> for AttributeMacro {
    fn format(
        &self,
        (attr, input): (proc_macro2::TokenStream, proc_macro2::TokenStream),
    ) -> String {
        let attr_ident = &self.0;
        let tokens = if attr.is_empty() {
            quote::quote! {
                #[#attr_ident]
                #input
            }
        } else {
            quote::quote! {
                #[#attr_ident(#attr)]
                #input
            }
        };
        format_input(tokens)
    }
}

fn format_input(input: proc_macro2::TokenStream) -> String {
    let input_string = syn::parse2(input.clone())
        .as_ref()
        .map(prettyplease::unparse)
        .unwrap_or_else(|_| input.to_string());
    input_string
}

#[track_caller]
fn expand_with<Fn: ProcMacroFn>(
    function: Fn,
    input: Fn::Input,
    attribute: impl AttributeFormatter<Fn::Input>,
    snapshot_name: &str,
) {
    let input_string = attribute.format(input.clone());
    let out = function.call(input);

    let file = syn::parse2(out).unwrap();

    let out = prettyplease::unparse(&file);
    let mut settings = insta::Settings::new();
    let content = vec![(
        insta::internals::Content::String("input".into()),
        insta::internals::Content::String(input_string),
    )];

    settings.set_raw_info(&insta::internals::Content::Map(content));

    let _scope = settings.bind_to_scope();

    insta::assert_snapshot!(snapshot_name, out);
}

mod as_changeset;
mod as_expression;
mod associations;
mod auto_type;
mod declare_sql_function;
mod define_sql_function;
mod diesel_for_each_tuple;
mod diesel_numeric_ops;
mod diesel_public_if;
mod from_sql_row;
mod has_query;
mod identifiable;
mod insertable;
mod multiconnection;
mod query_id;
mod queryable;
mod queryable_by_name;
mod selectable;
mod sql_function;
mod sql_type;
mod table;
mod valid_grouping;
