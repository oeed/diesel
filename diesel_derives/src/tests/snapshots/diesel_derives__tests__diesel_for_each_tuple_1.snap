---
source: diesel_derives/src/tests/mod.rs
expression: out
info:
  input: "__diesel_for_each_tuple! {\n    tuple_impls\n}\n"
---
tuple_impls! {
    1i32 { (0) -> T, ST, TT, } 2i32 { (0) -> T, ST, TT, (1) -> T1, ST1, TT1, } 3i32 { (0)
    -> T, ST, TT, (1) -> T1, ST1, TT1, (2) -> T2, ST2, TT2, } 4i32 { (0) -> T, ST, TT,
    (1) -> T1, ST1, TT1, (2) -> T2, ST2, TT2, (3) -> T3, ST3, TT3, } 5i32 { (0) -> T, ST,
    TT, (1) -> T1, ST1, TT1, (2) -> T2, ST2, TT2, (3) -> T3, ST3, TT3, (4) -> T4, ST4,
    TT4, } 6i32 { (0) -> T, ST, TT, (1) -> T1, ST1, TT1, (2) -> T2, ST2, TT2, (3) -> T3,
    ST3, TT3, (4) -> T4, ST4, TT4, (5) -> T5, ST5, TT5, } 7i32 { (0) -> T, ST, TT, (1) ->
    T1, ST1, TT1, (2) -> T2, ST2, TT2, (3) -> T3, ST3, TT3, (4) -> T4, ST4, TT4, (5) ->
    T5, ST5, TT5, (6) -> T6, ST6, TT6, } 8i32 { (0) -> T, ST, TT, (1) -> T1, ST1, TT1,
    (2) -> T2, ST2, TT2, (3) -> T3, ST3, TT3, (4) -> T4, ST4, TT4, (5) -> T5, ST5, TT5,
    (6) -> T6, ST6, TT6, (7) -> T7, ST7, TT7, } 9i32 { (0) -> T, ST, TT, (1) -> T1, ST1,
    TT1, (2) -> T2, ST2, TT2, (3) -> T3, ST3, TT3, (4) -> T4, ST4, TT4, (5) -> T5, ST5,
    TT5, (6) -> T6, ST6, TT6, (7) -> T7, ST7, TT7, (8) -> T8, ST8, TT8, } 10i32 { (0) ->
    T, ST, TT, (1) -> T1, ST1, TT1, (2) -> T2, ST2, TT2, (3) -> T3, ST3, TT3, (4) -> T4,
    ST4, TT4, (5) -> T5, ST5, TT5, (6) -> T6, ST6, TT6, (7) -> T7, ST7, TT7, (8) -> T8,
    ST8, TT8, (9) -> T9, ST9, TT9, } 11i32 { (0) -> T, ST, TT, (1) -> T1, ST1, TT1, (2)
    -> T2, ST2, TT2, (3) -> T3, ST3, TT3, (4) -> T4, ST4, TT4, (5) -> T5, ST5, TT5, (6)
    -> T6, ST6, TT6, (7) -> T7, ST7, TT7, (8) -> T8, ST8, TT8, (9) -> T9, ST9, TT9, (10)
    -> T10, ST10, TT10, } 12i32 { (0) -> T, ST, TT, (1) -> T1, ST1, TT1, (2) -> T2, ST2,
    TT2, (3) -> T3, ST3, TT3, (4) -> T4, ST4, TT4, (5) -> T5, ST5, TT5, (6) -> T6, ST6,
    TT6, (7) -> T7, ST7, TT7, (8) -> T8, ST8, TT8, (9) -> T9, ST9, TT9, (10) -> T10,
    ST10, TT10, (11) -> T11, ST11, TT11, } 13i32 { (0) -> T, ST, TT, (1) -> T1, ST1, TT1,
    (2) -> T2, ST2, TT2, (3) -> T3, ST3, TT3, (4) -> T4, ST4, TT4, (5) -> T5, ST5, TT5,
    (6) -> T6, ST6, TT6, (7) -> T7, ST7, TT7, (8) -> T8, ST8, TT8, (9) -> T9, ST9, TT9,
    (10) -> T10, ST10, TT10, (11) -> T11, ST11, TT11, (12) -> T12, ST12, TT12, } 14i32 {
    (0) -> T, ST, TT, (1) -> T1, ST1, TT1, (2) -> T2, ST2, TT2, (3) -> T3, ST3, TT3, (4)
    -> T4, ST4, TT4, (5) -> T5, ST5, TT5, (6) -> T6, ST6, TT6, (7) -> T7, ST7, TT7, (8)
    -> T8, ST8, TT8, (9) -> T9, ST9, TT9, (10) -> T10, ST10, TT10, (11) -> T11, ST11,
    TT11, (12) -> T12, ST12, TT12, (13) -> T13, ST13, TT13, } 15i32 { (0) -> T, ST, TT,
    (1) -> T1, ST1, TT1, (2) -> T2, ST2, TT2, (3) -> T3, ST3, TT3, (4) -> T4, ST4, TT4,
    (5) -> T5, ST5, TT5, (6) -> T6, ST6, TT6, (7) -> T7, ST7, TT7, (8) -> T8, ST8, TT8,
    (9) -> T9, ST9, TT9, (10) -> T10, ST10, TT10, (11) -> T11, ST11, TT11, (12) -> T12,
    ST12, TT12, (13) -> T13, ST13, TT13, (14) -> T14, ST14, TT14, } 16i32 { (0) -> T, ST,
    TT, (1) -> T1, ST1, TT1, (2) -> T2, ST2, TT2, (3) -> T3, ST3, TT3, (4) -> T4, ST4,
    TT4, (5) -> T5, ST5, TT5, (6) -> T6, ST6, TT6, (7) -> T7, ST7, TT7, (8) -> T8, ST8,
    TT8, (9) -> T9, ST9, TT9, (10) -> T10, ST10, TT10, (11) -> T11, ST11, TT11, (12) ->
    T12, ST12, TT12, (13) -> T13, ST13, TT13, (14) -> T14, ST14, TT14, (15) -> T15, ST15,
    TT15, } 17i32 { (0) -> T, ST, TT, (1) -> T1, ST1, TT1, (2) -> T2, ST2, TT2, (3) ->
    T3, ST3, TT3, (4) -> T4, ST4, TT4, (5) -> T5, ST5, TT5, (6) -> T6, ST6, TT6, (7) ->
    T7, ST7, TT7, (8) -> T8, ST8, TT8, (9) -> T9, ST9, TT9, (10) -> T10, ST10, TT10, (11)
    -> T11, ST11, TT11, (12) -> T12, ST12, TT12, (13) -> T13, ST13, TT13, (14) -> T14,
    ST14, TT14, (15) -> T15, ST15, TT15, (16) -> T16, ST16, TT16, } 18i32 { (0) -> T, ST,
    TT, (1) -> T1, ST1, TT1, (2) -> T2, ST2, TT2, (3) -> T3, ST3, TT3, (4) -> T4, ST4,
    TT4, (5) -> T5, ST5, TT5, (6) -> T6, ST6, TT6, (7) -> T7, ST7, TT7, (8) -> T8, ST8,
    TT8, (9) -> T9, ST9, TT9, (10) -> T10, ST10, TT10, (11) -> T11, ST11, TT11, (12) ->
    T12, ST12, TT12, (13) -> T13, ST13, TT13, (14) -> T14, ST14, TT14, (15) -> T15, ST15,
    TT15, (16) -> T16, ST16, TT16, (17) -> T17, ST17, TT17, } 19i32 { (0) -> T, ST, TT,
    (1) -> T1, ST1, TT1, (2) -> T2, ST2, TT2, (3) -> T3, ST3, TT3, (4) -> T4, ST4, TT4,
    (5) -> T5, ST5, TT5, (6) -> T6, ST6, TT6, (7) -> T7, ST7, TT7, (8) -> T8, ST8, TT8,
    (9) -> T9, ST9, TT9, (10) -> T10, ST10, TT10, (11) -> T11, ST11, TT11, (12) -> T12,
    ST12, TT12, (13) -> T13, ST13, TT13, (14) -> T14, ST14, TT14, (15) -> T15, ST15,
    TT15, (16) -> T16, ST16, TT16, (17) -> T17, ST17, TT17, (18) -> T18, ST18, TT18, }
    20i32 { (0) -> T, ST, TT, (1) -> T1, ST1, TT1, (2) -> T2, ST2, TT2, (3) -> T3, ST3,
    TT3, (4) -> T4, ST4, TT4, (5) -> T5, ST5, TT5, (6) -> T6, ST6, TT6, (7) -> T7, ST7,
    TT7, (8) -> T8, ST8, TT8, (9) -> T9, ST9, TT9, (10) -> T10, ST10, TT10, (11) -> T11,
    ST11, TT11, (12) -> T12, ST12, TT12, (13) -> T13, ST13, TT13, (14) -> T14, ST14,
    TT14, (15) -> T15, ST15, TT15, (16) -> T16, ST16, TT16, (17) -> T17, ST17, TT17, (18)
    -> T18, ST18, TT18, (19) -> T19, ST19, TT19, } 21i32 { (0) -> T, ST, TT, (1) -> T1,
    ST1, TT1, (2) -> T2, ST2, TT2, (3) -> T3, ST3, TT3, (4) -> T4, ST4, TT4, (5) -> T5,
    ST5, TT5, (6) -> T6, ST6, TT6, (7) -> T7, ST7, TT7, (8) -> T8, ST8, TT8, (9) -> T9,
    ST9, TT9, (10) -> T10, ST10, TT10, (11) -> T11, ST11, TT11, (12) -> T12, ST12, TT12,
    (13) -> T13, ST13, TT13, (14) -> T14, ST14, TT14, (15) -> T15, ST15, TT15, (16) ->
    T16, ST16, TT16, (17) -> T17, ST17, TT17, (18) -> T18, ST18, TT18, (19) -> T19, ST19,
    TT19, (20) -> T20, ST20, TT20, } 22i32 { (0) -> T, ST, TT, (1) -> T1, ST1, TT1, (2)
    -> T2, ST2, TT2, (3) -> T3, ST3, TT3, (4) -> T4, ST4, TT4, (5) -> T5, ST5, TT5, (6)
    -> T6, ST6, TT6, (7) -> T7, ST7, TT7, (8) -> T8, ST8, TT8, (9) -> T9, ST9, TT9, (10)
    -> T10, ST10, TT10, (11) -> T11, ST11, TT11, (12) -> T12, ST12, TT12, (13) -> T13,
    ST13, TT13, (14) -> T14, ST14, TT14, (15) -> T15, ST15, TT15, (16) -> T16, ST16,
    TT16, (17) -> T17, ST17, TT17, (18) -> T18, ST18, TT18, (19) -> T19, ST19, TT19, (20)
    -> T20, ST20, TT20, (21) -> T21, ST21, TT21, } 23i32 { (0) -> T, ST, TT, (1) -> T1,
    ST1, TT1, (2) -> T2, ST2, TT2, (3) -> T3, ST3, TT3, (4) -> T4, ST4, TT4, (5) -> T5,
    ST5, TT5, (6) -> T6, ST6, TT6, (7) -> T7, ST7, TT7, (8) -> T8, ST8, TT8, (9) -> T9,
    ST9, TT9, (10) -> T10, ST10, TT10, (11) -> T11, ST11, TT11, (12) -> T12, ST12, TT12,
    (13) -> T13, ST13, TT13, (14) -> T14, ST14, TT14, (15) -> T15, ST15, TT15, (16) ->
    T16, ST16, TT16, (17) -> T17, ST17, TT17, (18) -> T18, ST18, TT18, (19) -> T19, ST19,
    TT19, (20) -> T20, ST20, TT20, (21) -> T21, ST21, TT21, (22) -> T22, ST22, TT22, }
    24i32 { (0) -> T, ST, TT, (1) -> T1, ST1, TT1, (2) -> T2, ST2, TT2, (3) -> T3, ST3,
    TT3, (4) -> T4, ST4, TT4, (5) -> T5, ST5, TT5, (6) -> T6, ST6, TT6, (7) -> T7, ST7,
    TT7, (8) -> T8, ST8, TT8, (9) -> T9, ST9, TT9, (10) -> T10, ST10, TT10, (11) -> T11,
    ST11, TT11, (12) -> T12, ST12, TT12, (13) -> T13, ST13, TT13, (14) -> T14, ST14,
    TT14, (15) -> T15, ST15, TT15, (16) -> T16, ST16, TT16, (17) -> T17, ST17, TT17, (18)
    -> T18, ST18, TT18, (19) -> T19, ST19, TT19, (20) -> T20, ST20, TT20, (21) -> T21,
    ST21, TT21, (22) -> T22, ST22, TT22, (23) -> T23, ST23, TT23, } 25i32 { (0) -> T, ST,
    TT, (1) -> T1, ST1, TT1, (2) -> T2, ST2, TT2, (3) -> T3, ST3, TT3, (4) -> T4, ST4,
    TT4, (5) -> T5, ST5, TT5, (6) -> T6, ST6, TT6, (7) -> T7, ST7, TT7, (8) -> T8, ST8,
    TT8, (9) -> T9, ST9, TT9, (10) -> T10, ST10, TT10, (11) -> T11, ST11, TT11, (12) ->
    T12, ST12, TT12, (13) -> T13, ST13, TT13, (14) -> T14, ST14, TT14, (15) -> T15, ST15,
    TT15, (16) -> T16, ST16, TT16, (17) -> T17, ST17, TT17, (18) -> T18, ST18, TT18, (19)
    -> T19, ST19, TT19, (20) -> T20, ST20, TT20, (21) -> T21, ST21, TT21, (22) -> T22,
    ST22, TT22, (23) -> T23, ST23, TT23, (24) -> T24, ST24, TT24, } 26i32 { (0) -> T, ST,
    TT, (1) -> T1, ST1, TT1, (2) -> T2, ST2, TT2, (3) -> T3, ST3, TT3, (4) -> T4, ST4,
    TT4, (5) -> T5, ST5, TT5, (6) -> T6, ST6, TT6, (7) -> T7, ST7, TT7, (8) -> T8, ST8,
    TT8, (9) -> T9, ST9, TT9, (10) -> T10, ST10, TT10, (11) -> T11, ST11, TT11, (12) ->
    T12, ST12, TT12, (13) -> T13, ST13, TT13, (14) -> T14, ST14, TT14, (15) -> T15, ST15,
    TT15, (16) -> T16, ST16, TT16, (17) -> T17, ST17, TT17, (18) -> T18, ST18, TT18, (19)
    -> T19, ST19, TT19, (20) -> T20, ST20, TT20, (21) -> T21, ST21, TT21, (22) -> T22,
    ST22, TT22, (23) -> T23, ST23, TT23, (24) -> T24, ST24, TT24, (25) -> T25, ST25,
    TT25, } 27i32 { (0) -> T, ST, TT, (1) -> T1, ST1, TT1, (2) -> T2, ST2, TT2, (3) ->
    T3, ST3, TT3, (4) -> T4, ST4, TT4, (5) -> T5, ST5, TT5, (6) -> T6, ST6, TT6, (7) ->
    T7, ST7, TT7, (8) -> T8, ST8, TT8, (9) -> T9, ST9, TT9, (10) -> T10, ST10, TT10, (11)
    -> T11, ST11, TT11, (12) -> T12, ST12, TT12, (13) -> T13, ST13, TT13, (14) -> T14,
    ST14, TT14, (15) -> T15, ST15, TT15, (16) -> T16, ST16, TT16, (17) -> T17, ST17,
    TT17, (18) -> T18, ST18, TT18, (19) -> T19, ST19, TT19, (20) -> T20, ST20, TT20, (21)
    -> T21, ST21, TT21, (22) -> T22, ST22, TT22, (23) -> T23, ST23, TT23, (24) -> T24,
    ST24, TT24, (25) -> T25, ST25, TT25, (26) -> T26, ST26, TT26, } 28i32 { (0) -> T, ST,
    TT, (1) -> T1, ST1, TT1, (2) -> T2, ST2, TT2, (3) -> T3, ST3, TT3, (4) -> T4, ST4,
    TT4, (5) -> T5, ST5, TT5, (6) -> T6, ST6, TT6, (7) -> T7, ST7, TT7, (8) -> T8, ST8,
    TT8, (9) -> T9, ST9, TT9, (10) -> T10, ST10, TT10, (11) -> T11, ST11, TT11, (12) ->
    T12, ST12, TT12, (13) -> T13, ST13, TT13, (14) -> T14, ST14, TT14, (15) -> T15, ST15,
    TT15, (16) -> T16, ST16, TT16, (17) -> T17, ST17, TT17, (18) -> T18, ST18, TT18, (19)
    -> T19, ST19, TT19, (20) -> T20, ST20, TT20, (21) -> T21, ST21, TT21, (22) -> T22,
    ST22, TT22, (23) -> T23, ST23, TT23, (24) -> T24, ST24, TT24, (25) -> T25, ST25,
    TT25, (26) -> T26, ST26, TT26, (27) -> T27, ST27, TT27, } 29i32 { (0) -> T, ST, TT,
    (1) -> T1, ST1, TT1, (2) -> T2, ST2, TT2, (3) -> T3, ST3, TT3, (4) -> T4, ST4, TT4,
    (5) -> T5, ST5, TT5, (6) -> T6, ST6, TT6, (7) -> T7, ST7, TT7, (8) -> T8, ST8, TT8,
    (9) -> T9, ST9, TT9, (10) -> T10, ST10, TT10, (11) -> T11, ST11, TT11, (12) -> T12,
    ST12, TT12, (13) -> T13, ST13, TT13, (14) -> T14, ST14, TT14, (15) -> T15, ST15,
    TT15, (16) -> T16, ST16, TT16, (17) -> T17, ST17, TT17, (18) -> T18, ST18, TT18, (19)
    -> T19, ST19, TT19, (20) -> T20, ST20, TT20, (21) -> T21, ST21, TT21, (22) -> T22,
    ST22, TT22, (23) -> T23, ST23, TT23, (24) -> T24, ST24, TT24, (25) -> T25, ST25,
    TT25, (26) -> T26, ST26, TT26, (27) -> T27, ST27, TT27, (28) -> T28, ST28, TT28, }
    30i32 { (0) -> T, ST, TT, (1) -> T1, ST1, TT1, (2) -> T2, ST2, TT2, (3) -> T3, ST3,
    TT3, (4) -> T4, ST4, TT4, (5) -> T5, ST5, TT5, (6) -> T6, ST6, TT6, (7) -> T7, ST7,
    TT7, (8) -> T8, ST8, TT8, (9) -> T9, ST9, TT9, (10) -> T10, ST10, TT10, (11) -> T11,
    ST11, TT11, (12) -> T12, ST12, TT12, (13) -> T13, ST13, TT13, (14) -> T14, ST14,
    TT14, (15) -> T15, ST15, TT15, (16) -> T16, ST16, TT16, (17) -> T17, ST17, TT17, (18)
    -> T18, ST18, TT18, (19) -> T19, ST19, TT19, (20) -> T20, ST20, TT20, (21) -> T21,
    ST21, TT21, (22) -> T22, ST22, TT22, (23) -> T23, ST23, TT23, (24) -> T24, ST24,
    TT24, (25) -> T25, ST25, TT25, (26) -> T26, ST26, TT26, (27) -> T27, ST27, TT27, (28)
    -> T28, ST28, TT28, (29) -> T29, ST29, TT29, } 31i32 { (0) -> T, ST, TT, (1) -> T1,
    ST1, TT1, (2) -> T2, ST2, TT2, (3) -> T3, ST3, TT3, (4) -> T4, ST4, TT4, (5) -> T5,
    ST5, TT5, (6) -> T6, ST6, TT6, (7) -> T7, ST7, TT7, (8) -> T8, ST8, TT8, (9) -> T9,
    ST9, TT9, (10) -> T10, ST10, TT10, (11) -> T11, ST11, TT11, (12) -> T12, ST12, TT12,
    (13) -> T13, ST13, TT13, (14) -> T14, ST14, TT14, (15) -> T15, ST15, TT15, (16) ->
    T16, ST16, TT16, (17) -> T17, ST17, TT17, (18) -> T18, ST18, TT18, (19) -> T19, ST19,
    TT19, (20) -> T20, ST20, TT20, (21) -> T21, ST21, TT21, (22) -> T22, ST22, TT22, (23)
    -> T23, ST23, TT23, (24) -> T24, ST24, TT24, (25) -> T25, ST25, TT25, (26) -> T26,
    ST26, TT26, (27) -> T27, ST27, TT27, (28) -> T28, ST28, TT28, (29) -> T29, ST29,
    TT29, (30) -> T30, ST30, TT30, } 32i32 { (0) -> T, ST, TT, (1) -> T1, ST1, TT1, (2)
    -> T2, ST2, TT2, (3) -> T3, ST3, TT3, (4) -> T4, ST4, TT4, (5) -> T5, ST5, TT5, (6)
    -> T6, ST6, TT6, (7) -> T7, ST7, TT7, (8) -> T8, ST8, TT8, (9) -> T9, ST9, TT9, (10)
    -> T10, ST10, TT10, (11) -> T11, ST11, TT11, (12) -> T12, ST12, TT12, (13) -> T13,
    ST13, TT13, (14) -> T14, ST14, TT14, (15) -> T15, ST15, TT15, (16) -> T16, ST16,
    TT16, (17) -> T17, ST17, TT17, (18) -> T18, ST18, TT18, (19) -> T19, ST19, TT19, (20)
    -> T20, ST20, TT20, (21) -> T21, ST21, TT21, (22) -> T22, ST22, TT22, (23) -> T23,
    ST23, TT23, (24) -> T24, ST24, TT24, (25) -> T25, ST25, TT25, (26) -> T26, ST26,
    TT26, (27) -> T27, ST27, TT27, (28) -> T28, ST28, TT28, (29) -> T29, ST29, TT29, (30)
    -> T30, ST30, TT30, (31) -> T31, ST31, TT31, }
}
